src/Microsoft.OData.Core/HttpUtils.cs

```diff
@@ -453,37 +455,82 @@ internal static string ReadTokenOrQuotedStringValue(string headerName, string he
                         break;
                     }
 
+                    if (parameterValue == null)
+                    {
+                        parameterValue = new StringBuilder();

```

Comment:
I'm curious how using the string builder compares to allocating/renting a block large enough to store the encoded escaped string then using the string constructor with a slice at the end.

```c#
int maxLength = estimateUpperboundSizeOfParameterValue;

char[] array = null;
Span<char> parameterValue = maxLength < 128 ? stackalloc char[128] : array = ArrayPool<char>.Shared.Rent(maxLength);

while (...) {
  // copy text into parameterValue buffer instead of string builder.Append which can resize itself
}


// get string from span buffer:
return parameterValue.Slice(0, textLength).ToString();
if (array != null)
{
    ArrayPool<char>.Shared.Return(array);
}

Reply:
The difference here is that `StringBuilder.Append()` can allocate multiple arrays depending on the size of the string we append. The approach above proposes allocating a single buffer that is guaranteed to be large enough for the parameter value. We can estimate a minimal upper bound, it doesn't have to be the exact size. At the end we slice only the portion we've written to and create a string from that.

We'd need to compare both approaches to determine conclusively which is one more performant for the common sizes of parameterValue we expect to encounter.

The `stackalloc` and `ArrayPool` also help minimize allocations compared to `StringBuilder`.

Reply:
Header typically doesn't contain a long original string. So, I think The StringBuilder used here is enough. If it's critical issue for performance, we can file new issue to improve this part.

Reply:
#Resolved.